src/server/scripts/Outland/CoilfangReservoir/SerpentShrine/boss_leotheras_the_blind.cpp

```diff
@@ -80,267 +75,235 @@ const Position channelersPos[MAX_CHANNELERS] =
-                    events.ScheduleEvent(EVENT_SWITCH_TO_ELF, 60000);
-                    break;
-                case EVENT_SWITCH_TO_ELF:
-                    DoResetThreatList();
-                    me->LoadEquipment();
-                    me->GetMotionMaster()->MoveChase(me->GetVictim(), 0.0f);
-                    me->RemoveAurasDueToSpell(SPELL_METAMORPHOSIS);
-                    events.ScheduleEvent(EVENT_SWITCH_TO_DEMON, 55000);
-                    events.ScheduleEvent(EVENT_SPELL_WHIRLWIND, 10000);
-                    break;
-                case EVENT_SPELL_WHIRLWIND:
-                    me->CastSpell(me, SPELL_WHIRLWIND, false);
-                    events.ScheduleEvent(EVENT_SPELL_WHIRLWIND, 27000);
-                    break;
-                case EVENT_SPELL_INSIDIOUS_WHISPER:
-                    Talk(SAY_INNER_DEMONS);
-                    me->CastCustomSpell(SPELL_INSIDIOUS_WHISPER, SPELLVALUE_MAX_TARGETS, 5, me, false);
-                    break;
-                case EVENT_SUMMON_DEMON:
-                    me->CastSpell(me, SPELL_SUMMON_SHADOW_OF_LEOTHERAS, true);
-                    break;
-                case EVENT_RESTORE_FIGHT:
-                    me->SetStandState(UNIT_STAND_STATE_STAND);
-                    me->SetReactState(REACT_AGGRESSIVE);
-                    me->GetMotionMaster()->MoveChase(me->GetVictim());
-                    break;
-            }
+        scheduler.Update(diff);
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
 
-            if (me->GetDisplayId() == me->GetNativeDisplayId())
-                DoMeleeAttackIfReady();
-            else if (me->isAttackReady(BASE_ATTACK))
-            {
-                me->CastSpell(me->GetVictim(), SPELL_CHAOS_BLAST, false);
-                me->setAttackTimer(BASE_ATTACK, 2000);
-            }
+        if (me->GetDisplayId() == me->GetNativeDisplayId())
+        {
+            DoMeleeAttackIfReady();
         }
-    };
+        else if (me->isAttackReady(BASE_ATTACK))
+        {
+            me->CastSpell(me->GetVictim(), SPELL_CHAOS_BLAST, false);
+            me->setAttackTimer(BASE_ATTACK, 2000);
+        }
+    }
+private:
+    bool _recentlySpoken;
 };
 
-class npc_inner_demon : public CreatureScript
+struct npc_inner_demon : public ScriptedAI
 {
-public:
-    npc_inner_demon() : CreatureScript("npc_inner_demon") { }
-
-    CreatureAI* GetAI(Creature* creature) const override
+    npc_inner_demon(Creature* creature) : ScriptedAI(creature)
     {
-        return GetSerpentShrineAI<npc_inner_demonAI>(creature);
+        _scheduler.SetValidator([this]
+        {
+            return !me->HasUnitState(UNIT_STATE_CASTING);
+        });
+
+        instance = creature->GetInstanceScript();
     }
 
-    struct npc_inner_demonAI : public ScriptedAI
-    {
-        npc_inner_demonAI(Creature* creature) : ScriptedAI(creature)
-        {
-        }
+    InstanceScript* instance;
```

Comment:
move to private instead here. And also, before using this variable, always check for nullptr just in case.

Reply:
the register method ensures scripts are only ran inside the instance the creatured is registered into, so in practice instance won't ever be null

```cpp
RegisterSerpentShrineAI(npc_inner_demon);
```